tools/crm_resource_print.c

```diff
@@ -292,25 +291,35 @@ resource_search_list_default(pcmk__output_t *out, va_list args)
         return rc;
     }
 
-    for (GList *lpc = nodes; lpc != NULL; lpc = lpc->next) {
-        pe_node_t *node = (pe_node_t *) lpc->data;
+    for (GList *lpc1 = nodes; lpc1 != NULL; lpc1 = lpc1->next) {
+        pe_node_t *node = (pe_node_t *) lpc1->data;
 
-        if (!printed) {
-            out->begin_list(out, NULL, NULL, "Nodes");
-            printed = true;
-            rc = pcmk_rc_ok;
-        }
+        for (GList *lpc2 = node->details->running_rsc; lpc2 != NULL; lpc2 = lpc2->next) {
+            pe_resource_t *rsc = (pe_resource_t *) lpc2->data;
 
-        if (out->is_quiet(out)) {
-            out->list_item(out, "node", "%s", node->details->uname);
-        } else {
-            const char *state = "";
+            if (!pcmk__str_eq(requested_name, rsc->id, pcmk__str_none) &&
+                !pcmk__str_eq(requested_name, rsc->clone_name, pcmk__str_none) &&
+                !(pe_rsc_is_clone(rsc->parent) && pcmk__str_eq(requested_name, rsc->parent->id, pcmk__str_none))) {
+                continue;
+            }
```

Comment:
> This kind of feels like I am going around in a circle I don't need to be, but I'm not sure what I would do differently.

I puzzled over this myself while doing the bug investigation. It seems like any choice we make, we end up redoing some work.

Have you considered the following approach?
  - Pass in `rsc` like before
  - Get its `parent` (the clone wrapper)
  - For each `node`:
    - For each `child` in `parent`:
      - If `node` is in `child->running_on`, proceed with printing info for that resource.

I wonder if we might save some iterations and checks that way. I haven't fully thought this out.

Reply:
Another possibility would be for cli_resource_search() to create a list of not just nodes, but structs with node and a boolean for is-promoted. We could simplify by repurposing an existing pe_node_shared_s boolean (e.g. unpacked) for is-promoted, since we don't care about those values in this usage, to continue returning just a list of nodes at the expense of clarity. (Those booleans should really be a bitmask, in which case we could've just added a flag with no problem, but it's public API so we can't change it now.)

Reply:
I believe we're only really interested in node->details->uname, so it could be a list of struct { const char *node_name; const char *role; }. That would more cleanly separate the output code from the scheduler-aware code.

Reply:
"at the expense of clarity" makes me uneasy :)

I like the struct idea.

Reply:
or struct { const char *node_name; bool promoted; } ... you get the idea

Reply:
Okay, check that out.